source/server/http/config_tracker_impl.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <envoy/server/config_tracker.h>
+
+#include "common/common/assert.h"
+#include "common/common/macros.h"
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * Implementation of ConfigTracker.
+ */
+class ConfigTrackerImpl : public ConfigTracker {
+public:
+  EntryOwnerPtr add(const std::string& key, Cb cb) override;
+  const CbsMap& getCallbacksMap() const override;
+
+private:
+  std::shared_ptr<CbsMap> map_{std::make_shared<CbsMap>()};
+
+  class EntryOwnerImpl : public ConfigTracker::EntryOwner {
+  public:
+    EntryOwnerImpl(std::weak_ptr<CbsMap> map_weak, std::string key);
```

Comment:
Why is this only a `weak_ptr`? The map itself doesn't need to be weak does it? Style guide prefers simple ownership semantics where possible https://google.github.io/styleguide/cppguide.html#Ownership_and_Smart_Pointers.

Here's my strawman suggestion. Can we just assume ConfigTracker is going to outlive all EntryOwners? It should be effectively a singleton across the server, I think.

Reply:
Your second question answers your first - I wanted either to be able to outlive the other but for the map belong to ConfigTracker.

You're totally right with your strawman but as I intimated a few times in this PR I was planning on reusing ConfigTracker in situations where we couldn't guarantee it will effectively be a singleton.

Side note: +1 on simple ownership semantics, but that's exactly why I like to hide more complicated ownership situations behind simpler interfaces when they come up, as in this PR.

Reply:
In general my preference is to keep it simple unless the more complicated code is actually needed right now, but in the interest of moving this along I don't feel that strongly about it. Will defer to @htuch on it.

Either way, this PR needs a master merge now.

Reply:
@jsedgwick Ack. Looks like https://github.com/envoyproxy/envoy/pull/2771/files#diff-a112a0982616fac9024fc1c8e42f0661R175 is one such place, but the map is strongly owned there. If that's the only one, I'd say just change to shared_ptr for now and leave a TODO, that way we can see how it plays out.